### DISCUSSION TITLE
do I misunderstand the ConversionException in read_csv?

### DISCUSSION THREAD

**Discussion by eroell** [duckdb/duckdb#13510] [Q&A] [answered]
Hi y'all

new to duckdb and trying to use its Python API.

```py
import duckdb
duckdb.register("measurement_table", duckdb.read_csv( <insert path to csv> ))

duckdb.sql("SELECT * FROM measurement_table")
```

Yields `ConversionException`

```
ConversionException                       Traceback (most recent call last)
Cell In[59], [line 1](vscode-notebook-cell:?execution_count=59&line=1)
----> [1](vscode-notebook-cell:?execution_count=59&line=1) duckdb.sql("SELECT * FROM measurement_table").df()

ConversionException: Conversion Error: CSV Error on Line: 46000
Original Line: <some data>
Error when converting column "measurement_source_value". Could not convert string "HR [bpm]" to 'BIGINT'

Column measurement_source_value is being converted as type BIGINT
This type was auto-detected from the CSV file.
Possible solutions:
* Override the type for this column manually by setting the type explicitly, e.g. types={'measurement_source_value': 'VARCHAR'}
* Set the sample size to a larger value to enable the auto-detection to scan more values, e.g. sample_size=-1
* Use a COPY statement to automatically derive types from an existing table.

  file=
  delimiter = , (Set By User)
  quote = " (Set By User)
  escape = " (Set By User)
  new_line = \n (Auto-Detected)
  header = true (Set By User)
  skip_rows = 0 (Set By User)
  date_format =  (Auto-Detected)
  timestamp_format =  (Auto-Detected)
...
  sample_size=20480
  ignore_errors=false
  all_varchar=0
```

I spent a moment trying to find at which point I can add `types={'measurement_source_value': 'VARCHAR'}`, and actually still havent found it.

What works instead is `dtype={'measurement_source_value': 'VARCHAR'}` in `read_csv`.

```py
duckdb.register(
    "measurement_table",
    duckdb.read_csv(
        <insert path to csv>,
        dtype={"measurement_source_value": "VARCHAR"},
    ),
)
```

How would I go about using the instruction from the `ConversionError` string? Is the `ConversionError` string up to date? Could open an issue in the latter case, but I suspect its rather me missing something.

Thanks!

```py
pip show duckdb
```
```
Name: duckdb
Version: 1.0.0
```

**Comment by rpbouman** [comment, marked as answer]
Hi!

> I spent a moment trying to find at which point I can add types={'measurement_source_value': 'VARCHAR'}, and actually still haven't found it.

I'm not sure if I understand your question correctly. Do you want to know how to apply the advice coming out of the error message? The error message is produced by the CSV reader. In plain DuckDB SQL one can write:

`select * from read_csv('path/to/file', types={'measurement_source_value': 'VARCHAR'})`

> What works instead is dtype={'measurement_source_value': 'VARCHAR'} in read_csv.

According to https://duckdb.org/docs/data/csv/overview#parameters, `types` and `dtype` should be interchangeable, if you find that is not the case, consider filing an issue.

Alternatively, if all your columns are ok to be read as `varchar`, you could also consider passing `all_varchar=true`.

I hope this helps!

Best regards, 
Roland.

**Comment by eroell** [comment]
Thanks for the reply

>  Do you want to know how to apply the advice coming out of the error message? 

Yes.

Found that the [Python API reference ](https://duckdb.org/docs/api/python/reference/) correctly sticks to the `dtype` argument name, which is not in sync with the advice in the error and apparently duckdb SQL's `read_csv` function, but really minor enough

My confusion got resolved with this.
